src/index.ts

```diff
@@ -30,11 +27,72 @@ const transact = (doc: Y.Doc | null, opts: Options, fn: () => void) => {
   }
 };
 
+const toYValue = (val: any) => {
+  if (isProxyArray(val)) {
+    const arr = new Y.Array();
+    arr.insert(
+      0,
+      val.map(toYValue).filter((v) => v !== undefined && v !== null),
+    );
+    return arr;
+  }
+
+  if (isProxyObject(val)) {
+    const map = new Y.Map();
+    Object.entries(val).forEach(([key, value]) => {
+      const v = toYValue(value);
+      if (v !== undefined) {
+        map.set(key, v);
+      }
+    });
+    return map;
+  }
+
+  if (isPrimitiveMapValue(val)) {
+    return val;
+  }
+
+  return undefined;
+};
+
+const toJSON = (yv: unknown) => {
+  if (yv instanceof Y.AbstractType) {
+    return yv.toJSON();
+  }
+
+  return yv;
+};
+
+const getNestedValues = <T>(
+  p: Record<string, T> | T[],
+  y: Y.Map<T> | Y.Array<T>,
+  path: (string | number)[],
+) => {
+  let pv: any = p;
+  let yv: any = y;
+  for (let i = 0; i < path.length; i += 1) {
+    const k = path[i];
+    if (yv instanceof Y.Map) {
+      pv = pv[k];
+      yv = yv.get(k as string);
+    } else if (yv instanceof Y.Array) {
+      const index = Number(k);
+      pv = pv[k];
+      yv = yv.get(index);
+    } else {
+      pv = null;
+      yv = null;
+    }
+  }
+
+  return { p: pv, y: yv };
+};
+
```

Comment:
I like that you extracted some functions to the top. Do you have any opinions on code readability in this library in general? 
I initially had a hard time following what is happening exactly and am trying to address that with more elaborate variable names and by extracting many nested operations into their own functions that can be documented.

Reply:
I agree, there can be more done in this area. I think getting rid of the recursion helps a lot in terms of readability, but for this PR, I was attempting to conform to the existing style as much as possible.

Reply:
Yeah that makes sense.